packages/langium/src/grammar/references/langium-grammar-references.ts

```diff
@@ -0,0 +1,182 @@
+import { findNodeForFeature } from '../grammar-util';
+import { collectChildrenTypes, collectSuperTypes } from '../type-system/types-util';
+
+export class LangiumGrammarReferences extends DefaultReferences {
+    protected readonly documents: LangiumDocuments;
+
+    constructor(services: LangiumServices) {
+        super(services);
+        this.documents = services.shared.workspace.LangiumDocuments;
+    }
+
+    findDeclaration(sourceCstNode: CstNode): CstNode | undefined {
+        const nodeElem = findRelevantNode(sourceCstNode);
+        if (isAssignment(nodeElem)) {
+            return this.findAssignmentDeclaration(nodeElem);
+        }
+        return super.findDeclaration(sourceCstNode);
+    }
+
+    protected findLocalReferences(target: AstNode, includeDeclaration = false): Stream<ReferenceDescription> {
+        if (isTypeAttribute(target)) {
+            const doc = getDocument(target);
+            const rootNode = doc.parseResult.value;
+            return this.findLocalReferencesToTypeAttribute(target, rootNode);
+        } else {
+            return super.findLocalReferences(target, includeDeclaration);
+        }
+    }
+
+    protected findGlobalReferences(target: AstNode, includeDeclaration = false): Stream<ReferenceDescription> {
+        if (isTypeAttribute(target)) {
+            return this.findReferencesToTypeAttribute(target);
+        } else {
+            return super.findGlobalReferences(target, includeDeclaration);
+        }
+    }
+
+    protected findLocalReferencesToTypeAttribute(target: TypeAttribute, rootNode: AstNode): Stream<ReferenceDescription> {
+        const refs: ReferenceDescription[] = [];
+        const interfaceNode = getContainerOfType(target, isInterface);
+        if (interfaceNode) {
+            const interfaces = collectChildrenTypes(interfaceNode, this, this.documents, this.nodeLocator);
+            const targetRules: Array<ParserRule | Action> = [];
+            interfaces.forEach(interf => {
+                const rules = this.findLocalRulesWithReturnType(interf, rootNode);
+                targetRules.push(...rules);
+            });
+            if (equalURI(getDocument(target).uri, getDocument(rootNode).uri)) {
+                const ref = this.getReferenceToSelf(target);
+                if (ref) {
+                    refs.push(ref);
+                }
+            }
+            targetRules.forEach(rule => {
+                const assignment = isParserRule(rule) ?
+                    extractAssignments(rule.definition).find(a => a.feature === target.name)
+                    : getContainerOfType(rule, isGroup)!.elements.find(el => isAssignment(el) && el.feature === target.name);
+                if (assignment?.$cstNode) {
+                    const leaf = findLeafNodeAtOffset(assignment.$cstNode!, assignment.$cstNode!.offset);
```

Comment:
Suggestion: This doesn't return correct results. It does when we try to find the node that belongs to the `feature` property of the assignment.
```suggestion
                    const leaf = findNodeForFeature(assignment.$cstNode, 'feature');
```